AI/VCAI/Goals/Explore.cpp

```diff
@@ -179,3 +168,222 @@ bool Explore::fulfillsMe(TSubgoal goal)
 	}
 	return false;
 }
+
+bool Explore::hasReachableNeighbor(const int3 &pos, HeroPtr hero, CCallback * cbp, VCAI * vcai) const
+{
+	for(crint3 dir : int3::getDirs())
+	{
+		int3 tile = pos + dir;
+		if(cbp->isInTheMap(tile))
+		{
+			auto isAccessible = vcai->ah->isTileAccessible(hero, tile);
+
+			if(isAccessible)
+				return true;
+		}
+	}
+
+	return false;
+}
+
+int Explore::howManyTilesWillBeDiscovered(
+	const int3 & pos, 
+	int radious, 
+	CCallback * cbp, 
+	const TeamState * ts,
+	VCAI * aip,
+	HeroPtr h) const
+{
+	int ret = 0;
+	for(int x = pos.x - radious; x <= pos.x + radious; x++)
+	{
+		for(int y = pos.y - radious; y <= pos.y + radious; y++)
+		{
+			int3 npos = int3(x, y, pos.z);
+			if(cbp->isInTheMap(npos) 
+				&& pos.dist2d(npos) - 0.5 < radious 
+				&& !ts->fogOfWarMap[npos.x][npos.y][npos.z]
+				&& hasReachableNeighbor(npos, h, cbp, aip))
+			{
+				ret++;
+			}
+		}
+	}
+
+	return ret;
+}
+
+TSubgoal Explore::explorationBestNeighbour(int3 hpos, int radius, HeroPtr h) const
+{
+	std::map<int3, int> dstToRevealedTiles;
+	VCAI * aip = ai.get();
+	CCallback * cbp = cb.get();
+	const TeamState * ts = cbp->getPlayerTeam(ai->playerID);
+
+	for(crint3 dir : int3::getDirs())
+	{
+		int3 tile = hpos + dir;
+		if(cb->isInTheMap(tile))
+		{
+			if(isBlockVisitObj(tile))
+				continue;
+
+			if(isSafeToVisit(h, tile) && ai->isAccessibleForHero(tile, h))
+			{
+				auto distance = hpos.dist2d(tile); // diagonal movement opens more tiles but spends more mp
+				int tilesDiscovered = howManyTilesWillBeDiscovered(tile, radius, cbp, ts, aip, h);
+
+				dstToRevealedTiles[tile] = tilesDiscovered / distance;
+			}
+		}
+	}
+
+	if(dstToRevealedTiles.empty()) //yes, it DID happen!
+		return sptr(Invalid());
+
+	auto best = dstToRevealedTiles.begin();
+	for(auto i = dstToRevealedTiles.begin(); i != dstToRevealedTiles.end(); i++)
+	{
+		const CGPathNode * pn = cb->getPathsInfo(h.get())->getPathInfo(i->first);
+		//const TerrainTile *t = cb->getTile(i->first);
+		if(best->second < i->second && pn->reachable() && pn->accessible == CGPathNode::ACCESSIBLE)
+			best = i;
+	}
+
+	if(best->second)
+		return sptr(VisitTile(best->first).sethero(h));
+
+	return sptr(Invalid());
+}
+
+TSubgoal Explore::explorationNewPoint(HeroPtr h) const
+{
+	int radius = h->getSightRadius();
+	CCallback * cbp = cb.get();
+	VCAI *aip = ai.get();
+	const TeamState * ts = cbp->getPlayerTeam(aip->playerID);
+
+	std::vector<std::vector<int3>> tiles; //tiles[distance_to_fow]
+	tiles.resize(radius);
+
+	foreach_tile_pos([&](const int3 & pos)
+	{
+		if(!cbp->isVisible(pos))
+			tiles[0].push_back(pos);
```

Comment:
`tiles[0]` would contain all hidden tiles of entire map, you should at least reserve some space in a vector.